code/iaas/allocator/src/main/java/io/cattle/platform/allocator/dao/impl/AllocatorDaoImpl.java

```diff
@@ -486,4 +494,57 @@ public boolean hostHasContainerLabel(long hostId, String labelKey,
         }
         return instances;
     }
+    
+    /* (non-Javadoc)
+     * @see io.cattle.platform.allocator.dao.AllocatorDao#updateInstancePorts(java.util.Map)
+     * Scheduler will return a list of map showing the allocated result in the following format:
+     * {
+     *      "instanceID" : "xx",
+     *      "allocatedIPs" : {
+     *                              [
+     *                                  {
+     *                                      "allocatedIP" : "xxx.xxx.xxx.xxx",
+     *                                      "publicPort" : "xxx",
+     *                                      "privatePort" :  "xxx",
+     *                                  },
+     *                                  {
+     *                                      "allocatedIP" : "xxx.xxx.xxx.xxx",
+     *                                      "publicPort" : "xxx",
+     *                                      "privatePort" :  "xxx",
+     *                                  }
+     *                              ]
+     *                          }
+     * }
+     *      
+     * Then update 
+     * Port Table. Binding address field in port table and public port field in port table if public port is allocated by external scheduler (for agent)  
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public void updateInstancePorts(Map<String, Object> data) {
```

Comment:
To refactor, you can add a ports field to AlllocationAttempt. You can update it upon receiving the reserve response, but then move the logic for saving that to the instance into io.cattle.platform.allocator.dao.AllocatorDao.recordCandidate(AllocationAttempt, AllocationCandidate)